### PR TITLE
Fix NonBacktracking correctness issues and exponential blowup on nested loops

### PR DESCRIPTION
**Pull request by danmoseley**
## Fix NonBacktracking correctness issues and exponential blowup on nested loops

Fixes https://github.com/dotnet/runtime/issues/84188

### Problem

The NonBacktracking regex engine has exponential time complexity on deeply nested loop patterns like `((a)*)*`. The root cause is in the Brzozowski derivative computation: when the left side of a `Concat` is nullable (as with `*` loops), the derivative creates a 2-branch `Alternate`. With N levels of nesting, this produces O(2^N) intermediate nodes per derivative step.

| Nesting depth | Input length | Time (before) |
|---------------|-------------|---------------|
| 15 | 1000 | 1.62s |
| 17 | 1000 | 5.93s |
| 20 | 1000 | >60s |
| 2000 | 1 | 10.7s |

The last row shows that even matching a **single character** takes 10.7s at depth 2000 --- the bottleneck is the first derivative computation, not cumulative per-character cost.

### Fix

Two complementary changes:

**1. Loop flattening in `CreateLoop`** --- Simplify `(R*)*` to `R*` (and variants `(R+)*`, `(R*)+`) during symbolic AST construction. This directly eliminates the exponential state space. The simplification is restricted to cases where it preserves semantics:
- Inner lower bound must be 0 or 1 (with outer lower 0). Counterexample: `(R{2,})*` cannot match a single `R`, so must not become `R*`.
- Both loops must have the same laziness. Counterexample: `(?:R*)+?` (greedy inner, lazy outer) has different match behavior than `R*?`.
- No capture effects, since collapsing loops would change capture group bindings.

**2. `StripEffects` caching** --- The derivative of deeply nested loop patterns with captures produces a DAG with shared sub-trees wrapped in different `Effect` nodes. Without caching, `StripEffects` traverses each shared sub-tree once per reference rather than once per unique node, leading to exponential work. Adding a cache (following the existing pattern of `_pruneLowerPriorityThanNullabilityCache`) fixes this.

### Note on linear-time guarantee

The NonBacktracking engine guarantees linear time in the length of the input, and this guarantee was never violated by the nested loop issue. The engine is O(C * N) where N is the input length and C is a pattern-dependent constant representing the per-character cost. For deeply nested patterns like `((a)*)*`, the constant C grows exponentially with the nesting depth of the pattern, but for any given pattern the matching time remains linear in the input length.

In other words, this was an exponential blowup in the _pattern complexity constant_, not in input-length scaling. In practice, triggering the blowup requires deliberately nesting redundant quantifiers 15+ levels deep, which does not arise naturally.

This fix eliminates the blowup for same-laziness nesting without captures (via loop flattening) and for capturing patterns (via StripEffects caching). A theoretical gap remains for pathological non-capturing patterns that alternate lazy/greedy at every nesting level (e.g., `(?:(?:(?:a*?)*)*?)*` nested 20+ deep), since the loop flattening requires matching laziness. This is not a practical concern: such patterns would have to be deliberately constructed, and the engine already documents that patterns must be trusted.

### Tests

- Re-enables `StressTestDeepNestingOfLoops` for NonBacktracking at depth 120 (previously disabled citing #84188; even depth 20 was intractable before the fix).
- Adds 7 new correctness test cases for nested loop simplification edge cases, verified to fail without the fix and pass with it:
  - `^(?:(?:a){2,})*$` on `"a"` --- 0 matches (would be 1 if `body._lower >= 2` check were missing)
  - `(?:(?:a*)+?)` on `"aaa"` --- 2 matches (would be 4 if laziness guard were missing)
  - Sanity checks for patterns that are correctly simplified

### CODE REVIEW

**Review by olsaarik — COMMENTED** [review]
Overall the fix looks sounds and addresses a class of patterns that look like they should work without issue.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs [hunk 1]
```diff
@@ -416,6 +416,31 @@ internal static SymbolicRegexNode<TSet> CreateLoop(SymbolicRegexBuilder<TSet> bu
...
+            // Additional requirements:
+            //  - No effects (captures), since collapsing loops would change capture group bindings.
+            //  - Same laziness for both loops, since mixing greedy/lazy changes match priorities
+            //    (e.g. (?:0*)+? is not equivalent to 0*? — the former prefers fewer outer iterations
+            //    each greedily consuming, while the latter prefers less overall).
+            // This is a while loop (not if + recursive CreateLoop call) so that arbitrary nesting
+            // depth is handled without consuming stack proportional to the depth.
+            while (upper == int.MaxValue && body._kind == SymbolicRegexNodeKind.Loop && body._upper == int.MaxValue
+                && (body._lower == 0 || (body._lower == 1 && lower == 0))
+                && !body._info.ContainsEffect && isLazy == body.IsLazy)
```
[the remark this answers is not in the document]
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs]: AI has this response (I didn't create this PR in dotnet/runtime at this point because AI did all of this so far..). Thoughts?

====
The loop flattening does not apply to (((((a)*)*)*)*). Here's why:

   - When (a) is parsed, RegexNodeConverter wraps the body with CaptureStart/CaptureEnd Effect nodes before CreateLoop
   is called
   - So body._info.ContainsEffect == true, and the guard correctly blocks simplification
   - The stress test at depth 120 works because of the second fix (StripEffects caching), not loop flattening

  The loop flattening only applies to non-capturing patterns like (?:(?:a)*)*.
